metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorFactoryDecorator.java

```diff
@@ -1,81 +1,86 @@
 package com.netflix.metacat.common.server.connectors;
 
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
 import com.netflix.metacat.common.server.api.ratelimiter.RateLimiter;
 import lombok.Getter;
-import lombok.RequiredArgsConstructor;
+import lombok.NonNull;
 import lombok.extern.slf4j.Slf4j;
 
 /**
  * A decorator for a connector factory to add additional cross-cutting functionality
  * to all connector services.
  */
 @Slf4j
-@RequiredArgsConstructor
 public class ConnectorFactoryDecorator implements ConnectorFactory {
-    private final ConnectorPlugin connectorPlugin;
     @Getter
     private final ConnectorFactory delegate;
-    private final ConnectorContext connectorContext;
-    private final RateLimiter rateLimiter;
-
-    @Override
-    public ConnectorCatalogService getCatalogService() {
-        ConnectorCatalogService service = delegate.getCatalogService();
+    private final Supplier<ConnectorCatalogService> catalogService;
+    private final Supplier<ConnectorDatabaseService> databaseService;
+    private final Supplier<ConnectorTableService> tableService;
+    private final Supplier<ConnectorPartitionService> partitionService;
+
+    /**
+     * Creates the decorated connector factory that wraps connector services
+     * with additional wrappers.
+     *
+     * @param connectorPlugin the underlying plugin
+     * @param connectorContext the connector context for the underlying plugin
+     */
+    public ConnectorFactoryDecorator(@NonNull final ConnectorPlugin connectorPlugin,
+                                     @NonNull final ConnectorContext connectorContext) {
+        this.delegate = connectorPlugin.create(connectorContext);
+
+        if (isRateLimiterEnabled(connectorContext)) {
+            log.info("Creating rate-limited connector services for connector-type: {}, "
+                         + "plugin-type: {}, catalog: {}, shard: {}",
+                connectorContext.getConnectorType(), connectorPlugin.getType(),
+                connectorContext.getCatalogName(), connectorContext.getCatalogShardName());
 
-        if (isRateLimiterEnabled()) {
-            log.info("Creating rate-limited connector catalog services for connector-type: {}, "
+            final RateLimiter rateLimiter = connectorContext.getApplicationContext().getBean(RateLimiter.class);
+
+            // not all connectors implement these methods;
+            // calling these eagerly may throw an unsupported operation exception;
+            // hence we only memoize these for now and call them at runtime when they are needed
+            catalogService = memoize(
```

Comment:
Is the memoize mainly for delaying the eager call, if so could it cause issues due to returning cached result instead of new object each time

Reply:
Yeah, the memoization was to delay the eager call and should not cause issues because these services are singleton; but after sleeping on it I am thinking it is better not to memoize it and introduce that implicit coupling. Will move back to calling the delegate for every call and wrapping it with the throttling connector.